ceremonies/src/lib.rs

```diff
@@ -358,147 +375,465 @@ decl_error! {
+			<EndorseeCount>::insert((cid, cindex), 0);
+
+			<NewbieRegistry<T>>::remove_prefix((cid, cindex), None);
+			<NewbieIndex<T>>::remove_prefix((cid, cindex), None);
+			<NewbieCount>::insert((cid, cindex), 0);
+
+			<AssignmentCounts>::insert((cid, cindex), AssignmentCount::default());
+
+			Assignments::remove((cid, cindex));
+
 			<Endorsees<T>>::remove_prefix((cid, cindex), None);
-			<MeetupRegistry<T>>::remove_prefix((cid, cindex), None);
-			<MeetupLocationIndex<T>>::remove_prefix((cid, cindex), None);
 			<MeetupCount>::insert((cid, cindex), 0);
 			<AttestationRegistry<T>>::remove_prefix((cid, cindex), None);
 			<AttestationIndex<T>>::remove_prefix((cid, cindex), None);
 			<AttestationCount>::insert((cid, cindex), 0);
 			<MeetupParticipantCountVote<T>>::remove_prefix((cid, cindex), None);
+
+			<IssuedRewards>::remove_prefix((cid, cindex), None);
 		}
 		debug!(target: LOG, "purged registry for ceremony {}", cindex);
 	}
 
-	// this function is expensive, so it should later be processed off-chain within SubstraTEE-worker
-	// currently the complexity is O(n) where n is the number of registered participants
-	fn assign_meetups() {
-		let cids = <encointer_communities::Module<T>>::community_identifiers();
-		let cindex = <encointer_scheduler::Module<T>>::current_ceremony_index();
+	fn generate_meetup_assignment_params(community_ceremony: CommunityCeremony) -> DispatchResult {
+		let meetup_multiplier = 10u64;
+		let assignment_count =
+			Self::create_assignment_count(community_ceremony, meetup_multiplier)?;
+
+		let num_meetups =
+			ceil_division(assignment_count.get_number_of_participants(), meetup_multiplier);
+
+		Assignments::insert(
+			community_ceremony,
+			Assignment {
+				bootstrappers_reputables: Self::generate_assignment_function_params(
+					assignment_count.bootstrappers + assignment_count.reputables,
+					num_meetups,
+				),
+				endorsees: Self::generate_assignment_function_params(
+					assignment_count.endorsees,
+					num_meetups,
+				),
+				newbies: Self::generate_assignment_function_params(
+					assignment_count.newbies,
+					num_meetups,
+				),
+				locations: Self::generate_location_assignment_params(community_ceremony),
+			},
+		);
+
+		<AssignmentCounts>::insert(community_ceremony, assignment_count);
+		<MeetupCount>::insert(community_ceremony, num_meetups);
+		Ok(())
+	}
+
+	fn generate_location_assignment_params(
+		community_ceremony: CommunityCeremony,
+	) -> AssignmentParams {
+		let num_locations =
+			<encointer_communities::Module<T>>::get_locations(&community_ceremony.0).len() as u64;
+		AssignmentParams {
+			m: num_locations,
+			s1: Self::find_coprime_below(num_locations),
+			s2: Self::find_prime_below(num_locations),
+		}
+	}
+
+	fn find_coprime_below(upper_bound: u64) -> u64 {
+		let mut range: Vec<u64> = (1..upper_bound).collect();
+		range.shuffle(&mut StdRng::seed_from_u64(Self::get_random_nonzero_group_element(
+			u32::MAX.into(),
+		)));
```

Comment:
Why not just use the random permutation algorithm here, as used for the meetup assignments before:

```
    let mut random_source = RandomNumberGenerator::<T::Hashing>::new(
   // we don't need to pass a subject here, as this is only called once in a block.
   // However, without subject this should be initialized outside the community loop, otherwise
   // every community gets the same sequence of permutations if there are the same amount of
   // people per category (bootstrappers, reputables, etc.).
   T::RandomnessSource::random_seed().0,
);

range
   .random_permutation(&mut random_source)
   .unwrap_or_default()
   .into_iter(),
   .find_or_first(...)
...
```

However, this function is deterministic per block, so for the same upper_bound, repetitive calls to this will yield the same result. However, I believe this is not a problem as `find_coprime_below` is only called once per block. So I think a simple comment about that is fine.

Reply:
Fix no_std compilation in 3e127900df4603e491910e67c1ccc6ece816b61e